### PR TITLE
Implement save_graph for homogeneous graphs

### PR DESCRIPTION
**Pull request by jiaqima**
Add a helper function to help dataset contributors to save the data files and automatically create the `metadata.json` file.

## Description

Added an `io.py` to implement helper functions that help dataset contributors to create datasets in GLI formats. In this PR, we implemented a `save_graph` function to save .npz data files of the graph data and automatically create the `metadata.json` file. Currently only homogeneous graphs are supported. 

Also added an `Attribute` class to help guide contributors provide necessary information when adding an attribute to a node, an edge, or a graph. 

## Related Issue

This PR is partly adapted from PR #430 by @wood-ghost.

## Motivation and Context

Previously both metadata.json and task json files require the contributors to manually create, which is cumbersome. The new helper functions will significantly ease this process.

## How Has This Been Tested?

The implementation is tested on cora dataset by running the following snippet:

```python
# load and preprocess data
import numpy as np
import torch
import scipy.sparse as sparse
from dgl.data import CoraGraphDataset
dataset = CoraGraphDataset()
graph = dataset[0]

# cora_feat
node_feats = sparse.csr_matrix(graph.ndata["feat"].numpy())
# cora_class
node_class = graph.ndata["label"].numpy()  # (2708,)
# cora_edge
edge = torch.stack(graph.edges()).numpy().T
# cora only has 1 single connected graph
node_list = np.ones((1, graph.num_nodes()))  # (1, 2708)

# test save_graph
from gli.io import Attribute, save_graph

node_attr_feats = Attribute(
    name="NodeFeature",
    data=node_feats,
    description="Node features of Cora dataset, 1/0-valued vectors.")

node_attr_labels = Attribute(
    name="NodeLabel",
    data=node_class,
    description="Node labels of Cora dataset, int ranged from 1 to 7.")

save_graph(
    name="cora",
    edge=edge,
    node_attrs=[node_attr_feats, node_attr_labels],
    description="CORA dataset.",
    citation="@inproceedings{yang2016revisiting,\ntitle={Revisiting semi-supervised learning with graph embeddings},\nauthor={Yang, Zhilin and Cohen, William and Salakhudinov, Ruslan},\nbooktitle={International conference on machine learning},\npages={40--48},\nyear={2016},\norganization={PMLR}\n}"
)
```

The resulted json file is shown below:

```json
{
    "description": "CORA dataset.",
    "Node": {
        "NodeFeature": {
            "description": "Node features of Cora dataset, 1/0-valued vectors.",
            "type": "float",
            "format": "SparseTensor",
            "file": "cora__graph__Node_NodeFeature__7032c9c380d1889061dcbbcd76b8c427.sparse.npz"
        },
        "NodeLabel": {
            "description": "Node labels of Cora dataset, int ranged from 1 to 7.",
            "type": "int",
            "format": "Tensor",
            "file": "cora__graph__1d11e44bae2f3af9ad60bdb538a9a536.npz",
            "key": "Node_NodeLabel"
        }
    },
    "Edge": {
        "_Edge": {
            "file": "cora__graph__1d11e44bae2f3af9ad60bdb538a9a536.npz",
            "key": "Edge_Edge"
        }
    },
    "Graph": {
        "_NodeList": {
            "file": "cora__graph__1d11e44bae2f3af9ad60bdb538a9a536.npz",
            "key": "Graph_NodeList"
        }
    },
    "citation": "@inproceedings{yang2016revisiting,\ntitle={Revisiting semi-supervised learning with graph embeddings},\nauthor={Yang, Zhilin and Cohen, William and Salakhudinov, Ruslan},\nbooktitle={International conference on machine learning},\npages={40--48},\nyear={2016},\norganization={PMLR}\n}",
    "is_heterogeneous": false
}
```

As a comparison, this is the hand-written metadata.json in our current repo:
```json
{
    "description": "CORA dataset.",
    "data": {
        "Node": {
            "NodeFeature": {
                "description": "Node features of Cora dataset, 1/0-valued vectors.",
                "type": "int",
                "format": "SparseTensor",
                "file": "cora_node_feats.sparse.npz"
            },
            "NodeLabel": {
                "description": "Node labels of Cora dataset, int ranged from 1 to 7.",
                "type": "int",
                "format": "Tensor",
                "file": "cora.npz",
                "key": "node_class"
            }
        },
        "Edge": {
            "_Edge": {
                "file": "cora.npz",
                "key": "edge"
            }
        },
        "Graph": {
            "_NodeList": {
                "file": "cora.npz",
                "key": "node_list"
            },
            "_EdgeList": {
                "file": "cora.npz",
                "key": "edge_list"
            }
        }
    },
    "citation": "@inproceedings{yang2016revisiting,\ntitle={Revisiting semi-supervised learning with graph embeddings},\nauthor={Yang, Zhilin and Cohen, William and Salakhudinov, Ruslan},\nbooktitle={International conference on machine learning},\npages={40--48},\nyear={2016},\norganization={PMLR}\n}",
    "is_heterogeneous": false
}
```

### CODE REVIEW

**Comment by jiaqima** [comment]
@wood-ghost, please also take a look.

File: gli/io.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    metadata["is_heterogeneous"] = is_heterogeneous
+
+    if citation == "":
+        warnings.warn("The citation is empty.")
+    if is_heterogeneous:
+        raise NotImplementedError(
+            "Heterogeneous graphs are not supported yet.")
+
+    with open("metadata.json", "w", encoding="utf-8") as f:
+        json.dump(metadata, f, indent=4)
```
Comment (xingjian-zhang) [inline gli/io.py:237]: Should we use an optional `save_dir` keyword to specify the saving path of `metadata.json`?

**Review by xingjian-zhang — APPROVED** [review]
LGTM. The only issue is about docstring: We now rely on the [sphinx docstring format](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html) to build the doc page. I will try to use chatgpt or other tool to convert them directly to the sphinx format.

File: gli/io.py (hunk 1 above)
Comment (jiaqima) [reply gli/io.py:253]: Sounds good. Will add it.

Comment (jiaqima) [reply gli/io.py:253]: Actually, maybe not. The `gli.utils.save_data` does not have an dir argument. Saving npz files and `metadata.json` to different directories might be confusing?

Or maybe I can add save_dir to both save_data and here?

Comment (jiaqima) [reply gli/io.py]: Added save_dir to both in [3abd4ca](https://github.com/Graph-Learning-Benchmarks/gli/pull/431/commits/3abd4caf08a4a61311208d6eb49c45496040884a)

**Pull request merged**